pkg/csi-handler/pv_checker.go

```diff
@@ -0,0 +1,223 @@
+	if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != checker.checkerName {
+		return fmt.Errorf("csi source is nil or the volume is not managed by this checker/monitor")
+	}
+
+	if pv.Status.Phase != v1.VolumeBound {
+		return fmt.Errorf("PV: %s status is not bound", pv.Name)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), checker.timeout)
+	defer cancel()
+
+	volumeHandle, err := checker.GetVolumeHandle(pv)
+	if err != nil {
+		klog.Errorf("Get volume handler error: %+v", err)
+		return err
+	}
+
+	if len(volumeHandle) == 0 {
+		return fmt.Errorf("volume handle in csi source is empty")
+	}
+
+	volumeCondition, err := checker.csiPVHandler.ControllerVolumeChecking(ctx, volumeHandle)
+	if err != nil {
+		return err
+	}
+
+	// if error is not nil, the volumeCondition result can not be nil neither
+
+	// At the first stage, we just send PVC events
+	if volumeCondition.GetAbnormal() {
+		// Since pv status is bound, we believe PV controller, do not check pv.Spec.ClaimRef here.
+		pvc, err := checker.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+		if err != nil {
+			return err
+		}
+		checker.eventRecorder.Event(pvc, v1.EventTypeWarning, "PV is not healthy(detected by monitor controller)", volumeCondition.GetMessage())
+	}
+
+	return nil
+}
+
+// CheckNodeVolumeStatus checks volume status in node side
+func (checker *PVHealthConditionChecker) CheckNodeVolumeStatus(kubeletRootPath string, supportStageUnstage bool, pv *v1.PersistentVolume, pod *v1.Pod) error {
+	if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != checker.checkerName {
+		return fmt.Errorf("csi source is nil or the volume is not managed by this checker/monitor")
+	}
+
+	if pv.Status.Phase != v1.VolumeBound {
+		return fmt.Errorf("PV: %s status is not bound", pv.Name)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), checker.timeout)
+	defer cancel()
+
+	volumeHandle, err := checker.GetVolumeHandle(pv)
+	if err != nil {
+		klog.Errorf("Get volume handler error: %+v", err)
+		return err
+	}
+
+	if len(volumeHandle) == 0 {
+		return fmt.Errorf("volume handle in csi source is empty")
+	}
+
+	var volumePath, stagingTargetPath string
+
+	volumePath = util.GetVolumePath(kubeletRootPath, pv.Name, string(pod.UID))
+
+	if supportStageUnstage {
+		stagingTargetPath, err = util.MakeDeviceMountPath(kubeletRootPath, pv)
+		if err != nil {
+			return err
+		}
+	}
+
+	volumeCondition, err := checker.csiPVHandler.NodeVolumeChecking(ctx, volumeHandle, volumePath, stagingTargetPath)
+	if err != nil {
+		return err
+	}
+
+	// if error is not nil, the volumeCondition result can not be nil neither
+
+	// At the first stage, we just send PVC events
+	if volumeCondition.GetAbnormal() {
+		// Since pv status is bound, we believe PV controller, do not check pv.Spec.ClaimRef here.
+		pvc, err := checker.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
```

Comment:
During the CSI spec review, we discussed that it could be confusing when messages are coming from both Controller and Node and one says abnomal=True and the other one says abnormal=False.  There was a suggestion to have messages from Node to be added as an event on the Node, not on PVC.

Let me get some confirmation on that.

Reply:
Discussed this at the CSI Implementation meeting today.   Let's have Node plugin report events on Pod object.  If multiple pods are using the same volume, report events on all pods.

Reply:
>During the CSI spec review, we discussed that it could be confusing when messages are coming from both Controller and Node and one says abnomal=True and the other one says abnormal=False. 

If the status is normal, there will not be a event message here, so i don't think it is confusing

Anyway, i can add this to Pod